base_layer/core/src/covenants/covenant.rs

```diff
@@ -46,19 +49,32 @@ pub struct Covenant {
     tokens: Vec<CovenantToken>,
 }
 
+impl BorshSerialize for Covenant {
+    fn serialize<W: Write>(&self, writer: &mut W) -> io::Result<()> {
+        self.write_to(writer)
+    }
+}
+
+impl<'a> BorshDeserialize for Covenant {
+    fn deserialize(buf: &mut &[u8]) -> io::Result<Self> {
+        let covenant = Self::from_bytes(buf).map_err(|e| io::Error::new(io::ErrorKind::InvalidInput, e.to_string()))?;
```

Comment:
The covenant is variable length, so the length needs to be encoded 
Does this node work?
```rust
 fn deserialize(reader: &mut &[u8]) -> io::Result<Self> {
       let len = reader.read_varint::<usize>()?;
        if len == 0 {
            return Ok(Covenant::new());
        };
        // Check the length varint - this may be maliciously misreported
        if len > MAX_COVENANT_BYTES {
            return Err(io::Error::new(
                io::ErrorKind::InvalidInput,
                "covenants: Length varint exceeded maximum",
            ));
        }
        // Ensure that no more than the maximum bytes can be read
        let mut limited = LimitedBytesReader::new(MAX_COVENANT_BYTES, reader);
        CovenantTokenDecoder::new(&mut limited)
            .collect::<Result<_, CovenantDecodeError>>()
}
```

Reply:
How was it broken?

Reply:
In any case, happy with the changes for this :+1:

Reply:
> How was it broken?

Because the borsh will send you all data, so if the covenants were like couple bytes. The CovenantTokenDecoder received more than it shoulld. This is also why I added a tests where I send more data to the deserialization and I expect the buffer still contain the extra data.

Reply:
I should probably also refactor the test for extra data. Because now I have the same code in all 5-6 tests.